### ISSUE TITLE
detailed/extended environment creation/update

### ISSUE THREAD

**Issue by RonnyPfannschmidt** [mitsuhiko/pipsi#31]
after playing around with some ideas i came up with a scheme for more complex usage of pipsi

use cases examples are things like
- install tox and detox in the same env and get all commands
- install py.test with pytest-cache and pytest-xdist plugins
- install ipython with notebook support

this mainly involves installing more than one package with commands as well as installing plugins for packages with the commands

this creates the need to track a sets of details
1. list of packages/requirements to instal
2. list of packages whose commands to symlink
3. a list of extra pip arguments to propperly support #27 even for updates

i think the best way to support storing that is a json file in the virtualenv root

once them metadata is in place one can add extra commands on top of it

what i currently have in mind is
- create_env $env -- creates a empty venv
- set_pip_args $env -- configures the pip arguments for a venv
- install_package $env spec -- install a python package
- install_command $env spec -- maybe install a python package and link its scripts

once those details are in place, the update command can be altered to take all packages and the extra args into account

since this is a initial draft idea, please scorn and take appart

**Comment by untitaker** [comment]
Some things that I've encountered while testing [ronnys fork](https://github.com/RonnyPfannschmidt/pipsi/tree/win32-install) under Windows 8:

  - pipsi and it's bootstrapping script have the assumption that venvs contain "lib" and "bin" folders, while in my tests, it's "Scripts", "Lib" and "Include".
  - subprocess.call raises ``WindowsError`` on unknown commands
  - Apart from the commercial Anaconda distribution, there doesn't seem to be a sane Python installation for Windows that adds itself to path and comes with pip out-of-the-box.

**Comment by untitaker** [comment]
From IRC:

    14:10:34    untitaker | I think that functionality is out-of-scope for
                          | pip
    14:10:38    untitaker | *pipsi
    14:11:08    untitaker | I'd like pipsi to become a user-facing tool for
                          | people who want to install Python-based CLI
                          | tools without caring about Python's packaging
                          | mess
    14:11:10      Yatekii | ok now I just need to get git running in pycharm
    14:11:17      Yatekii | then I can dev VERY nicely!
    14:11:34    untitaker | I hope that I can stop telling users to just
                          | dump stuff into ~/.local/lib/
    14:11:38        ronny | untitaker: i want to do things like install tox
                          | and detox in the same env, or install pytest
                          | with plugins ^^
    14:12:01    untitaker | ronny: but you can already do that by manually
                          | activating the venvs from pipsi!
    14:12:23    untitaker | also it seems like you want to implement half of
                          | virtualenvmanager in pipsi
    14:12:41    untitaker | like I said, ideally the user wouldn't have to
                          | hear the word "virtualenv" IMO
    14:13:21        ronny | untitaker: so maybe add_commands/add_packages?

**Comment by untitaker** [comment]
Maybe add these options to ``pipsi install``:

 -  ``--allow-libraries``: Don't do the scripts check when installing PyPI packages, don't uninstall stuff when no scripts are found.
  - ``--use-venv=...``: Reuse a particular virtualenv from ``~/.local/venvs/`` for installation.

**Comment by RonnyPfannschmidt** [comment]
for the commands maybe a combination of the following can work
* `install --into $name $package` will install the given apckage into a preexisting venv and record the installed command
* `install --library-for $name $package` will install the given package into the env and record the package without copying the package

however in order to make it safe pipsi update still would need to know about the packages

**Comment by untitaker** [comment]
Then pipsi should track the packages that are explicitly installed (like OS package managers)

**Comment by RonnyPfannschmidt** [comment]
#52

**Comment by RonnyPfannschmidt** [comment]
closing to get if off my plate on github
